src/dodal/common/beamlines/instantiation_behaviour.py

```diff
@@ -0,0 +1,125 @@
+    config: DeviceInitializationConfig | None = None
+
+    def __init__(
+        self, config: DeviceInitializationConfig, factory: Callable[[], AnyDevice]
+    ) -> None:
+        self.factory = factory
+        self.config = config
+
+    # TODO right now the cache is in a global variable ACTIVE_DEVICES, that should change
+    def see_if_device_is_in_cache(self, name: str) -> AnyDevice | None:
+        d = ACTIVE_DEVICES.get(name)
+        assert isinstance(d, AnyDevice) or d is None
+        return d
+
+    def add_device_to_cache(self, device: AnyDevice) -> None:
+        ACTIVE_DEVICES[device.name] = device
+
+    def __call__(
+        self, connect=False, mock: bool | None = None, timeout: float | None = None
+    ) -> AnyDevice:
+        device: AnyDevice = (
+            # todo if we do not pass the name to the factory, we can not check if the device is in the cache.
+            # there are many devices from the same factory
+            self.see_if_device_is_in_cache(self.factory.__name__) or self.factory()
+        )
+        if self.config and self.config.set_name:
+            device.set_name(self.factory.__name__)
+        self.add_device_to_cache(device)
+
+        if connect:
+            assert self.config
+            call_in_bluesky_event_loop(
+                device.connect(
+                    timeout=self.config.default_timeout_for_connect
+                    if timeout is None
+                    else timeout,
+                    mock=self.config.default_use_mock_at_connection
+                    if mock is None
+                    else mock,
+                )
+            )
+
+        return device
+
+
+def instantiation_behaviour(
+    eager: bool = True,
+    set_name: bool = True,
+    default_timeout_for_connect: float = DEFAULT_TIMEOUT,
+    default_use_mock_at_connection: bool = False,
+) -> Callable[[Callable[[], AnyDevice]], DeviceInitializationController]:
+    config = DeviceInitializationConfig(
+        eager, set_name, default_timeout_for_connect, default_use_mock_at_connection
+    )
+
+    def decorator(factory: Callable[[], AnyDevice]) -> DeviceInitializationController:
+        controller = DeviceInitializationController(config, factory)
+        return controller
+
+    return decorator
+
+
+beamline_prefix = "example:"
+
+
+@instantiation_behaviour(
+    default_use_mock_at_connection=True, default_timeout_for_connect=10
+)
+def new_detector_xyz():
+    """Create an XYZ detector with specific settings."""
+    return XYZDetector(name="det1", prefix=f"{beamline_prefix}xyz:")
+
+
+@instantiation_behaviour(
+    eager=True, default_use_mock_at_connection=True, default_timeout_for_connect=10
+)
+def detector_xyz_variant():
+    """Create an XYZ detector with specific settings."""
+    return XYZDetector(name="det2-variant", prefix=f"{beamline_prefix}xyz:")
+
+
+cached_devices: dict[str, AnyDevice] = {
+    "det1": new_detector_xyz(),
+    "det2": detector_xyz_variant(),
+}
```

Comment:
What is this for?

Reply:
this had been for the controller, as before it was returning the controller.

This was here for demonstration purposes, I will delete this when making the PR ready

Reply:
RE: controller registering I'll pursue the second option, the first one sounds odd